[Tracking/Asahi] Upstream-ready Rust changes

hoshinolina opened the pull request:
Opening this tracking PR to document the Rust changes that are ready to be merged into `rust-next`.

fbq commented on rust/kernel/time.rs:
```diff
@@ -0,0 +1,23 @@
+//! C header: [`include/linux/timekeeping.h`](../../../../include/linux/timekeeping.h)
+
+use crate::bindings;
+use core::time::Duration;
```
`hrtimer` wrapper is in my TODO list, so it's nice to see this ;-)

`core::time::Duration` has a slight problem of being more than 64bit, we probably need to introduce our own `Duration` as the counterpart of `ktime_t`. But it's just my own opinion, time subsystem maintainers may have different ones.

I don't think it blocks this change, just want to mention it while we are at it.

bjorn3 commented on rust/macros/helpers.rs:
```diff
@@ -1,6 +1,6 @@
 // SPDX-License-Identifier: GPL-2.0
```
I believe the right tag is Co-developed-by and a sign-off is required from everyone who is listed as Co-developed-by.

hoshinolina replied:
I guess I can just copy the signoff from the original commits then? It's just that one simple macro which is already in the `rust` branch (I just followed the history to make sure I got everyone who touched it).

bjorn3 replied:
I believe so based on reading https://docs.kernel.org/process/submitting-patches.html#when-to-use-acked-by-cc-and-co-developed-by

hoshinolina replied:
Done!

bjorn3 commented on rust/kernel/print.rs:
```diff
@@ -147,12 +147,14 @@ macro_rules! print_macro (
+                    crate::__LOG_PREFIX,
+                    args,
+                );
+            }
```
This fix has been merged upstream in https://lore.kernel.org/rust-for-linux/167461338342.26446.17308850166282928589.pr-tracker-bot@kernel.org/T/#t already.

hoshinolina replied:
That's great! I'll drop this one then. Sorry for being so late to bringing this up (I noticed quite a while ago but it had been sitting in my branch...)

hoshinolina commented:
Updated:

- Rebased on latest `rust-for-next`
- Added some commits from asahi-soc at the top (these are queued for v6.3 and can be ignored)
- Imported a bunch of `error` module features from the `rust` branch.
- Added the XArray abstraction (now that `ForeignOwnable` is in)
- Added a stub `Device` abstraction based on the initial commit that introduced it to `rust`. This is enough to allow other parts of the `kernel` crate that interact with devices to be upstreamed, even though proper drivers will require more.
- Added the Apple RTKit abstraction (this is what depends on the asahi-soc patches). Although this normally depends on ARM64 support, it can be build-tested on amd64 with `CONFIG_COMPILE_TEST=y`.

hoshinolina commented:
Good news! With just a couple new bits, this branch is good enough to add most of the DRM abstractions my driver used as of last week, so I've gone ahead and added those. 🎉

hoshinolina commented:
@wedsonaf Can you sign off on c219c9a0ee / 6646d85a8a? That's the only `rust` branch commit I've seen so far that was missing the signoff that I need...

YakoYakoYokuYoku commented on rust/kernel/xarray.rs:
```diff
@@ -0,0 +1,296 @@
+// SPDX-License-Identifier: GPL-2.0
```
Conflicts with #884.

YakoYakoYokuYoku replied:
Actually this one can be ignored, I prefer to rebase my work on top of this.

fbq commented on rust/kernel/xarray.rs:
```diff
@@ -0,0 +1,296 @@
+/// # Invariants
+///
+/// The `*mut T` is always non-NULL and owned by the referenced `XArray`
+pub struct Guard<'a, T: ForeignOwnable>(*mut T, &'a Opaque<bindings::xarray>);
```
Maybe:
```suggestion
pub struct Guard<'a, T: ForeignOwnable>((NonNull<T>, &'a XAarray<T>);
```
?

And you can avoid the "Invariants" section

wedsonaf replied:
We need an unsafe constructor for this type. Otherwise one could create a guard to say 'NonNull::dangling', and then call `borrow`, which would deref the invalid pointer.

IOW, an invariants sections is not sufficient: we need to ensure that the invariants actually hold. If we can't do it with the type system, we need to make the operation unsafe and have the developer promise to abide by the requirements.

hoshinolina replied:
@fbq Fixed for the upstream submission!

@wedsonaf But the fields are private, nobody can construct it outside of this module ^^

fbq commented on rust/kernel/xarray.rs:
```diff
@@ -0,0 +1,296 @@
+            // Consider it a success anyway, not much we can do
+        }
+        let index = self.1;
+        core::mem::forget(self);
```
Maybe some comments for the `forget`?

hoshinolina replied:
Done for the upstream submission, thanks!

fbq commented on rust/kernel/xarray.rs:
```diff
@@ -0,0 +1,296 @@
+    fn deref(&self) -> &Self::Target {
+        // SAFETY: See the `borrow()` method. The dereferenced `T::Borrowed` value
+        // must share the same lifetime, so we can return a reference to it.
+        // TODO: Is this really sound?
```
Yeah, this is very interesting. I tried a few lifetime tricks, and couldn't make it work. @bjorn3 @nbdd0121 @wedsonaf anything you guys can think of? I think the intention here is to express that the lifetime of `Borrowed` outlives the lifetime of `&Self::Target`. But I may be wrong...

nbdd0121 replied:
This could be written as:
```rust
impl<'a, T: ForeignOwnable> Deref for Guard<'a, T>
where
    T::Borrowed<'a>: Deref,
    for<'b> T::Borrowed<'b>: Into<&'b <T::Borrowed<'a> as Deref>::Target>,
{
    type Target = <T::Borrowed<'a> as Deref>::Target;

    fn deref(&self) -> &Self::Target {
        self.borrow().into()
    }
}
```
then we just need to ensure `ArcBorrow<'a, T>: Into<'a T>` for this to be usable for `Arc`.

nbdd0121 replied:
Current code, as is, isn't really sound because `T::Borrowed<'self>: Deref` can do something funky inside and return a pointer to itself, rather than the thing it points to. What we really want here is a way to convert `T::Borrowed<'self>` to `&'self Target`, so it seems to me that just adding an explicit `Into` is the best way.

hoshinolina replied:
Yeah, that was my feeling when writing this too (that it's safe given *existing* `T::Borrowed` implementations that follow certain rules but not sound in the general case). I'll add an `Into` impl to `ArcBorrow` for the next revision and see how that goes, thanks! @nbdd0121 do you want to be added as `Co-developed-by`?

hoshinolina replied:
I couldn't figure this out... `Into` conflicts with the blanket `From` impl.

Since this is just a convenience, I think we can leave it for a future patch, so I've dropped this for now ^^;;

fbq commented on rust/kernel/xarray.rs:
```diff
@@ -0,0 +1,296 @@
+    ///
+    /// If `value` is `None`, then the index is reserved from further allocation but remains
+    /// free for storing a value into it.
+    pub fn alloc_limits(&self, value: Option<T>, min: u32, max: u32) -> Result<usize> {
```
Given we already have the `reserve_*` APIs, maybe better not providing another way for user to reserve slot?

How about a private `__alloc_limits`? And `alloc_*` APIs look like the below
```rust
fn __alloc_limits(&self, value: Option<T>, min: u32, max: u32) -> Result<usize>

pub fn alloc_limits(&self, value: T, min: u32, max: u32) -> Result<usize>
pub fn alloc(&self, value: T, min: u32, max: u32) -> Result<usize>
```

hoshinolina replied:
Done for the upstream submission, thanks!

bjorn3 commented on rust/kernel/xarray.rs:
```diff
@@ -0,0 +1,296 @@
+where
+    T::Borrowed<'static>: Deref,
+{
+    type Target = <T::Borrowed<'static> as Deref>::Target;
```
I think this will need to be `<T::Borrowed<'a> as Deref>::Target` as `T` is only guaranteed to live for `'a` and not for `'static`. Otherwise depending on the `ForeignOwnable` impl I think it may be possible to extract an `'static` reference out of it.

fbq replied:
I think so, but if we want use `self.borrow()` instead of `T::borrow()` in the `deref` implementation, how can we make it work?

bjorn3 replied:
No idea

nbdd0121 commented on rust/kernel/xarray.rs:
```diff
@@ -0,0 +1,296 @@
+    /// Returns a `T::Borrowed` type for the owned `ForeignOwnable` type.
+    pub fn borrow<'b>(&'b self) -> T::Borrowed<'b>
+    where
+        'a: 'b,
```
This bound is implied from the well-formedness of `&'b Self`

hoshinolina replied:
Removed it for the upstream submission, thanks! ^^

wedsonaf commented:
> @wedsonaf Can you sign off on [c219c9a](https://github.com/Rust-for-Linux/linux/commit/c219c9a0eef03e4a86f710daa050f1b430a4f5d2) / [6646d85](https://github.com/Rust-for-Linux/linux/commit/6646d85a8a72adee9a133139f673a9dc9cd006e7)? That's the only `rust` branch commit I've seen so far that was missing the signoff that I need...

Ah, this predates our practice of signing off commits. Anyway:

Signed-off-by: Wedson Almeida Filho <wedsonaf@gmail.com>

wedsonaf commented on rust/kernel/xarray.rs:
```diff
@@ -0,0 +1,296 @@
+
+/// Flag values passed to `XArray::new` to configure the `XArray`.
+pub mod flags {
+    /// Use IRQ-safe locking
```
nit: please add a period to the end of these comments (for consistency with other abstractions).

hoshinolina replied:
Done for the upstream submission, thanks!

wedsonaf commented on rust/kernel/xarray.rs:
```diff
@@ -0,0 +1,296 @@
+pub struct Reservation<'a, T: ForeignOwnable>(&'a XArray<T>, usize, PhantomData<T>);
+
+impl<'a, T: ForeignOwnable> Reservation<'a, T> {
+    /// Store a value into the reserved slot.
```
nit: please use the third-person conjugation here, i.e., "Stores ...".

hoshinolina replied:
Done for the upstream submission, thanks!

wedsonaf commented on rust/kernel/xarray.rs:
```diff
@@ -0,0 +1,296 @@
+        let old = unsafe {
+            bindings::xa_store(
+                self.xa.get(),
+                index.try_into()?,
```
You're leaking the new value here if `try_into` fails.

My suggestion is that you use a `ScopeGuard` that calls from `from_foreign(new)` and is dismissed on the success path.

hoshinolina replied:
Done for the upstreaming patch, thanks! ^^

wedsonaf commented on rust/kernel/xarray.rs:
```diff
@@ -0,0 +1,296 @@
+    }
+
+    /// Replaces an entry with a new value, returning the old value (if any).
+    pub fn replace(&self, index: usize, value: T) -> Result<Option<T>> {
```
Missing SAFETY comments in this function.

hoshinolina replied:
Added for the upstream submission, thanks!

wedsonaf reviewed (commented):
Leaving some comments on xarray abstraction

hoshinolina commented:
Since the DRM stuff will be reviewed on top of everything else, I've taken the part that was in this MR out (and added a few things). Still haven't addressed all the review feedback, I'll do that soon!

YakoYakoYokuYoku commented on rust/kernel/xarray.rs:
```diff
@@ -0,0 +1,299 @@
+
+impl<T: ForeignOwnable> XArray<T> {
+    /// Creates a new `XArray` with the given flags.
+    pub fn new(flags: Flags) -> Result<XArray<T>> {
```
Could this constructor be made constant? It would be pretty useful for static mutexes, instead of using a lazy value like `LazyCell`.

bjorn3 replied:
Unfortunately not. It calls the `xa_init_flags` foreign function. Const functions are only allowed to call other const functions defined in rust.

YakoYakoYokuYoku replied:
Then the way to make it happen would be to either inline the `xa_init_flags` function or have another constructor that declares an `xarray` in place, though either ways still require a raw `spinlock_t`.

https://github.com/Rust-for-Linux/linux/blob/1dd685c414a7b9fdb3d23aca3aedae84f0b998ae/include/linux/xarray.h#L296-L301

YakoYakoYokuYoku replied:
Wait, I have an idea.

```rust
    pub fn new(flags: Flags) -> Result<XArray<T>> {
        let xa = Opaque::uninit();

        // SAFETY: We have just created `xa`. This data structure does not require
        // pinning.
        unsafe {
            xa.get().write_bytes(0, 1);
            (*xa.get()).xa_flags = flags;
        }

        // INVARIANT: Initialize the `XArray` with a valid `xa`.
        XArray {
            xa,
            _p: PhantomData,
        }
    }
```

fbq replied:
`spin_lock_init()` is more complicated then `write_bytes(0, 1)`, but I guess making it `const` is doable.

hoshinolina replied:
I don't think I'm really comfortable with this, with all the spinlock variations involved (RT builds, lockdep, etc.)... I don't think the upstream maintainers of all that will want to duplicate all of that logic in Rust...

fbq replied:
Oh, I wasn't talking about upstreaming this, just it's doable, which can give us some ideas in the future how much we can gain if we make XA new `const`.

YakoYakoYokuYoku commented on rust/kernel/xarray.rs:
```diff
@@ -0,0 +1,299 @@
+        unsafe { bindings::xa_init_flags(xa.get(), flags) };
+
+        // INVARIANT: Initialize the `XArray` with a valid `xa`.
+        Ok(XArray {
```
As this is always infallible I see no reason to return a `Result<XArray<T>>`, `XArray<T>` is a nobrainer.

hoshinolina replied:
Good catch, fixed for the upstream submission! ^^

YakoYakoYokuYoku reviewed (commented):
Actually, I have a pair of notes on the constructor of `XArray<T>`, I've encountered some issues while using it.

ojeda commented on rust/kernel/sync/arc.rs:
```diff
@@ -489,6 +510,14 @@ impl<T> UniqueArc<MaybeUninit<T>> {
+        unsafe { self.assume_init() }
+    }
+
+    /// Returns a UniqueArc<T>, assuming the MaybeUninit<T> has already been initialized.
+    ///
+    /// # Safety
+    /// The contents of the UniqueArc must have already been fully initialized.
```
```suggestion
    /// Returns a `UniqueArc<T>`, assuming the `MaybeUninit<T>` has already been initialized.
    ///
    /// # Safety
    ///
    /// The contents of the `UniqueArc<T>` must have already been fully initialized.
```

i.e. blank line after a section header, Markdown wherever possible and being consistent in the second `UniqueArc` mention.

Also please try to see if the intra-doc links work, e.g.

```rust
[`MaybeUninit<T>`]
```

or

```rust
[`MaybeUninit`]
```

ojeda commented on rust/kernel/xarray.rs:
```diff
@@ -0,0 +1,299 @@
+    pub const LOCK_IRQ: super::Flags = bindings::BINDINGS_XA_FLAGS_LOCK_IRQ;
+    /// Use softirq-safe locking
+    pub const LOCK_BH: super::Flags = bindings::BINDINGS_XA_FLAGS_LOCK_BH;
+    /// Track which entries are free (distinct from None)
+    pub const TRACK_FREE: super::Flags = bindings::BINDINGS_XA_FLAGS_TRACK_FREE;
+    /// Initialize array index 0 as busy
+    pub const ZERO_BUSY: super::Flags = bindings::BINDINGS_XA_FLAGS_ZERO_BUSY;
```
```suggestion
    /// Track which entries are free (distinct from None)
    pub const TRACK_FREE: super::Flags = bindings::BINDINGS_XA_FLAGS_TRACK_FREE;

    /// Initialize array index 0 as busy
    pub const ZERO_BUSY: super::Flags = bindings::BINDINGS_XA_FLAGS_ZERO_BUSY;
```

i.e. blank line between when there are docs (e.g. `file.rs`).

ojeda commented on rust/kernel/xarray.rs:
```diff
@@ -0,0 +1,299 @@
+///
+/// # Invariants
+///
+/// The `*mut T` is always non-NULL and owned by the referenced `XArray`
```
```suggestion
/// The `*mut T` is always non-null and owned by the referenced `XArray`.
```

i.e. period at the end & lowercase "non-null". Also try to see if intra-doc links work.

ojeda commented on rust/kernel/xarray.rs:
```diff
@@ -0,0 +1,299 @@
+        'a: 'b,
+    {
+        // SAFETY: The value is owned by the `XArray`, the lifetime it is borrowed for must not
+        // outlive the `XArray` itself, nor the Guard that holds the lock ensuring the value
```
```suggestion
        // outlive the `XArray` itself, nor the `Guard` that holds the lock ensuring the value
```

ojeda commented on rust/kernel/xarray.rs:
```diff
@@ -0,0 +1,299 @@
+    }
+}
+
+// Convenience impl for `ForeignOwnable` types whose `Borrowed`
+// form implements Deref.
```
```suggestion
// Convenience `impl` for `ForeignOwnable` types whose `Borrowed`
// form implements `Deref`.
```

ojeda commented on rust/kernel/xarray.rs:
```diff
@@ -0,0 +1,299 @@
+            T::from_foreign(new);
+        });
+
+        let old = unsafe {
```
Missing `SAFETY` comment.

ojeda commented on rust/kernel/xarray.rs:
```diff
@@ -0,0 +1,299 @@
+    /// Replaces an entry with a new value, returning the old value (if any).
+    pub fn replace(&self, index: usize, value: T) -> Result<Option<T>> {
+        let new = value.into_foreign();
+        let guard = ScopeGuard::new(|| unsafe {
```
Missing `SAFETY` comment.

ojeda commented on rust/kernel/xarray.rs:
```diff
@@ -0,0 +1,299 @@
+    /// Allocate a new index in the array, optionally storing a new value into it, with
+    /// configurable bounds for the index range to allocate from.
+    ///
+    /// If `value` is `None`, then the index is reserved from further allocation but remains
```
```suggestion
    /// If `value` is [`None`], then the index is reserved from further allocation but remains
```

ojeda commented on rust/kernel/xarray.rs:
```diff
@@ -0,0 +1,299 @@
+
+    /// Reserve a new index in the array within configurable bounds for the index.
+    ///
+    /// Returns a `Reservation` object, which can then be used to store a value at this index or
```
```suggestion
    /// Returns a [`Reservation`] object, which can then be used to store a value at this index or
```

(Assuming it works)

ojeda commented on rust/kernel/xarray.rs:
```diff
@@ -0,0 +1,299 @@
+    }
+}
+
+// SAFETY: XArray is thread-safe and all mutation operations are internally locked.
```
```suggestion
// SAFETY: `XArray` is thread-safe and all mutation operations are internally locked.
```

ojeda commented on rust/kernel/soc/apple/rtkit.rs:
```diff
@@ -0,0 +1,277 @@
+    }
+}
+
+// SAFETY: `RtKit` operations require a mutable reference
```
```suggestion
// SAFETY: `RtKit` operations require a mutable reference.
```

ojeda commented on rust/kernel/soc/mod.rs:
```diff
@@ -0,0 +1,5 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! SoC drivers
```
```suggestion
//! SoC drivers.
```

ojeda commented on rust/kernel/soc/apple/mod.rs:
```diff
@@ -0,0 +1,6 @@
+// SPDX-License-Identifier: GPL-2.0-only OR MIT
+
+//! Apple SoC drivers
```
```suggestion
//! Apple SoC drivers.
```

ojeda commented on rust/kernel/soc/apple/rtkit.rs:
```diff
@@ -0,0 +1,277 @@
+///
+/// # Invariants
+///
+/// The rtk pointer is valid.
+/// The data pointer is a valid pointer from T::Data::into_foreign().
```
```suggestion
/// The `rtk` pointer is valid.
/// The `data` pointer is a valid pointer from `T::Data::into_foreign()`.
```

ojeda commented on rust/kernel/soc/apple/rtkit.rs:
```diff
@@ -0,0 +1,277 @@
+    }
+}
+
+/// Represents `struct apple_rtkit *`.
```
```suggestion
/// Represents the C side `struct apple_rtkit *`.
```

or:

```suggestion
/// Represents C's `struct apple_rtkit *`.
```

We don't have a single style (sometimes we also say "kernel's" too), and I know there was a `clk.rs` one like that, but I think eventually it would be nice to mention at least "C" (or alternative solutions I am considering for cross-doc linking).

ojeda commented on rust/kernel/soc/apple/rtkit.rs:
```diff
@@ -0,0 +1,277 @@
+    bfr: *mut bindings::apple_rtkit_shmem,
+) {
+    let bfr_mut = unsafe { &mut *bfr };
+    // SAFETY: Per shmem_setup_callback, this has to be a pointer to a Buffer if it is set.
+    if !bfr_mut.private.is_null() {
+        unsafe {
```
This `SAFETY` comment should be moved near the `unsafe`, also Markdown.

```suggestion
    if !bfr_mut.private.is_null() {
        // SAFETY: Per `shmem_setup_callback`, this has to be a pointer to a `Buffer` if it is set.
        unsafe {
```

ojeda commented on rust/kernel/ioctl.rs:
```diff
@@ -0,0 +1,64 @@
+// SPDX-License-Identifier: GPL-2.0
+#![allow(non_snake_case)]
+
+//! ioctl() number definitions
```
```suggestion
//! `ioctl()` number definitions.
```

Why the parenthesis, by the way?

ojeda commented on rust/kernel/io_pgtable.rs:
```diff
@@ -0,0 +1,347 @@
+                ))
+            }
+
+            /// Get the configuration for this IOPagetable.
```
```suggestion
            /// Get the configuration for this `IOPagetable`.
```

Maybe intra-doc link too.

Similarly nearby.

ojeda commented on rust/kernel/io_pgtable.rs:
```diff
@@ -0,0 +1,347 @@
+            coherent_walk: config.coherent_walk,
+            tlb: &Self::FLUSH_OPS,
+            iommu_dev: dev.raw_device(),
+            __bindgen_anon_1: unsafe { mem::zeroed() },
```
Missing `SAFETY` comment (several in this file).

ojeda commented on rust/kernel/io_pgtable.rs:
```diff
@@ -0,0 +1,347 @@
+        Self: Sized;
+}
+
+/// A generic IOMMU page table
```
```suggestion
/// A generic IOMMU page table.
```

ojeda commented on rust/kernel/io_pgtable.rs:
```diff
@@ -0,0 +1,347 @@
+    );
+}
+
+/// Inner page table info shared across all table types.
+/// # Invariants
```
```suggestion
/// Inner page table info shared across all table types.
///
/// # Invariants
```

ojeda commented on rust/kernel/io_pgtable.rs:
```diff
@@ -0,0 +1,347 @@
+
+    /// Optional callback to queue up leaf TLB invalidation for a single page.
+    ///
+    /// IOMMUs that cannot batch TLB invalidation operations efficiently will typically issue
+    /// them here, but others may decide to update the iommu_iotlb_gather structure and defer
+    /// the invalidation until iommu_iotlb_sync() instead.
```
```suggestion
    /// IOMMUs that cannot batch TLB invalidation operations efficiently will typically issue
    /// them here, but others may decide to update the `iommu_iotlb_gather` structure and defer
    /// the invalidation until `iommu_iotlb_sync()` instead.
```

ojeda commented on rust/kernel/io_pgtable.rs:
```diff
@@ -0,0 +1,347 @@
+
+/// Represents a requested io_pgtable configuration.
+pub struct Config {
+    /// Quirk bitmask (type-specific).
+    pub quirks: usize,
```
Blank lines (several groups in this file).

ojeda commented on rust/kernel/io_pgtable.rs:
```diff
@@ -0,0 +1,347 @@
+    pub const PRIV: u32 = bindings::IOMMU_PRIV;
+}
+
+/// Represents a requested io_pgtable configuration.
```
```suggestion
/// Represents a requested `io_pgtable` configuration.
```

ojeda commented on rust/kernel/io_pgtable.rs:
```diff
@@ -0,0 +1,347 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! IOMMU page table management
```
```suggestion
//! IOMMU page table management.
```

ojeda reviewed (commented):
Some nits here and there for your v2.
